kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,952 @@
+	}
+}
+
+func (c *loadBalancerController) acceptScheduledRequests(stopCh <-chan struct{}) {
+	log.Infof("Will run periodic IPAM sync every %s", timer)
+	t := time.NewTicker(timer)
+	for {
+		select {
+		case update := <-c.syncerUpdates:
+			c.handleUpdate(update)
+		case <-t.C:
+			log.Infof("Running periodic IPAM sync of Service LoadBalancer")
+			c.syncIPAM()
+		case <-c.syncChan:
+			c.syncIPAM()
+		case svcObj := <-c.serviceUpdates:
+			err := c.syncService(svcObj)
+			if err != nil {
+				log.WithError(err).Error("Error syncing service object, will retry during next IPAM sync")
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (c *loadBalancerController) handleUpdate(update interface{}) {
+	switch update := update.(type) {
+	case bapi.SyncStatus:
+		c.syncStatus = update
+		switch update {
+		case bapi.InSync:
+			log.WithField("status", update).Info("Syncer is InSync, kicking sync channel")
+			kick(c.syncChan)
+		}
+		return
+	case model.KVPair:
+		switch update.Key.(type) {
+		case model.ResourceKey:
+			switch update.Key.(model.ResourceKey).Kind {
+			case api.KindIPPool:
+				c.handleIPPoolUpdate(update)
+				kick(c.syncChan)
+				return
+			}
+		case model.BlockKey:
+			c.handleBlockUpdate(update)
+			return
+		}
+	}
+}
+
+func (c *loadBalancerController) handleBlockUpdate(kvp model.KVPair) {
+	if kvp.Value != nil {
+		affinity := kvp.Value.(*model.AllocationBlock).Affinity
+
+		if affinity != nil && *affinity == fmt.Sprintf("%s:%s", ipam.AffinityTypeVirtual, api.VirtualLoadBalancer) {
+			c.ipamBlocks[kvp.Key.String()] = kvp
+		}
+	} else {
+		delete(c.ipamBlocks, kvp.Key.String())
+	}
+}
+
+func (c *loadBalancerController) handleIPPoolUpdate(kvp model.KVPair) {
+	if kvp.Value != nil {
+		pool := kvp.Value.(*api.IPPool)
+		if slices.Contains(pool.Spec.AllowedUses, api.IPPoolAllowedUseLoadBalancer) {
+			c.ipPools[kvp.Key.String()] = *pool
+		}
+	} else {
+		delete(c.ipPools, kvp.Key.String())
+	}
+}
+
+// syncIpam has two main uses. It functions as a garbage collection for leaked IP addresses from Service LoadBalancer
+// The other use case is to update IPs for any Service LoadBalancer that do not have IPs assigned, this could be caused by the user
+// creating Service LoadBalancer before any valid pools were created
+func (c *loadBalancerController) syncIPAM() {
+	svcObjs, err := c.getServiceObjectList()
+	if err != nil {
+		return
+	}
+
+	for _, svcObj := range svcObjs {
+		err = c.syncService(svcObj)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to sync service %s/%s", svcObj.service.Namespace, svcObj.service.Name)
+			continue
+		}
+	}
+
+	for _, block := range c.ipamBlocks {
+		b := block.Value.(*model.AllocationBlock)
+		for key := range b.SequenceNumberForAllocation {
+			ordinal, err := strconv.Atoi(key)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to parse Ordinal for block %s", key)
+			}
+			ip := b.OrdinalToIP(ordinal)
+
+			if _, ok := c.allocationTracker.servicesByIp[ip.String()]; !ok {
```

Comment:
At least according to how I was imagining this would work, the `allocationTracker` was to be a reflection of the data from the IPAM blocks. We already have the Service data available to us in the Service indexer / lister. So, comparing `c.ipamBlocks` with `c.allocationTracker` isn't a useful comparison. I see that right now you've used `c.allocationTracker` more as a reflection of the `Status` field, but I worry (as mentioned elsewhere) that the Status field is an unreliable source of truth for which IP addresses are actually allocated in the cluster.

So, I would expect this code to iterate the IPAM data, use the metadata to reverse engineer the Service NamespacedName, and then compare the IPAM data to the actual Service data (or the allocation tracker, which should be equivalent).

One way to do that might be something like this? (needs a bit more thought)

```go
// Track all services to sync across the k8s API and our IPAM data base.
servicesToCheck := map[string]bool{}

// Add services we know about from the API.
for _, svc := range serviceLister.List() {
  servicesToCheck[namespacedName(svc)] = true
}

// Add services we know about from our IPAM data, if not present in the API.
for svc := range c.allocationTracker.ipsByService {
  servicesToCheck[namespacedName(svc)] = true
}

// Sync all of the services we know about (whether they exist or not). 
// syncService handles cleaning up any IPs for the service if it doesn't exist, and allocates / releases IPs 
// as needed for those that do.
for svc := servicesToCheck {
  syncService(svc)
}
```